### PR TITLE
Use priority queue for decoys instead of AVL tree

### PR DESCRIPTION
**Pull request by david415**
DO NOT MERGE

### CODE REVIEW

File: server/internal/decoy/decoy.go [hunk 1]
```diff
@@ -72,11 +72,16 @@ type decoy struct {
...
 
-	surbETAs   *avl.Tree
+	surbETAs   *queue.PriorityQueue
 	surbStore  map[uint64]*surbCtx
 	surbIDBase uint64
 }
 
+func addPriority(ctx *surbCtx) {
+	now := monotime.Now()
+	ctx.priority = uint64(now + ctx.eta)
```
Comment (mixmasala) [inline server/internal/decoy/decoy.go:82]: can we ensure that priority is always unique?

File: server/internal/decoy/decoy.go [hunk 2]
```diff
@@ -385,30 +379,7 @@ func (d *decoy) loadAndDeleteSURBCtx(id uint64) *surbCtx {
...
-				return ctx
-			}
-		}
-		panic("BUG: SURB missing from node ETA list")
-	}
-
-	d.surbETAs.Remove(ctx.etaNode)
-	ctx.etaNode = nil
-
+	d.surbETAs.RemovePriority(ctx.priority)
```
Comment (mixmasala) [inline server/internal/decoy/decoy.go:382]: check return value of RemovePriority != nil ( see panic("BUG.."))

**Review by mixmasala — CHANGES_REQUESTED** [review]
I think this is probabalistically safe as our PriorityQueue may not handle items with the same priority appropriately; it would also be nice to see a comparison benchmark between the implementations.

File: server/internal/decoy/decoy.go (hunk 1 above)
Comment (david415) [reply server/internal/decoy/decoy.go:82]: we don't need to ensure it's unique

File: server/internal/decoy/decoy.go [hunk 3]
```diff
@@ -365,14 +365,8 @@ func (d *decoy) storeSURBCtx(ctx *surbCtx) {
...
-	ctxList := []*surbCtx{ctx}
-	ctx.etaNode = d.surbETAs.Insert(ctxList)
-	if nCtxList := ctx.etaNode.Value.([]*surbCtx); nCtxList[0] != ctx {
-		// Existing node, append the context to the list stored
-		// as the node value.  This should be exceedingly rare.
-		ctx.etaNode.Value = append(nCtxList, ctx)
-	}
-
+	addPriority(ctx)
+	d.surbETAs.Enqueue(ctx.priority, ctx)
```
[the remark this answers is not in the document]
Comment (david415) [reply server/internal/decoy/decoy.go:369]: no it doesn't loose the reference. i just now wrote a test for this.

File: server/internal/decoy/decoy.go (hunk 2 above)
Comment (david415) [reply server/internal/decoy/decoy.go:382]: check the return value for what!?

Comment (david415) [reply server/internal/decoy/decoy.go:382]: this cannot crash. try it.

File: server/internal/decoy/decoy.go [hunk 4]
```diff
@@ -423,29 +394,19 @@ func (d *decoy) sweepSURBCtxs() {
...
 			break
 		}
-
-		for _, ctx := range surbCtxs {
-			delete(d.surbStore, ctx.id)
-			// TODO: At some point, this should do more than just log.
-			d.log.Debugf("Sweep: Lost SURB ID: 0x%08x ETA: %v (DeltaT: %v)", ctx.id, ctx.eta, now-ctx.eta)
-			swept++
+		ctx := entry.Value.(*surbCtx)
+		now := monotime.Now()
```
[the remark this answers is not in the document]
Comment (david415) [reply server/internal/decoy/decoy.go:404]: fixed

**Comment by david415** [comment]
> I think this is probabalistically safe as our PriorityQueue may not handle items with the same priority appropriately; it would also be nice to see a comparison benchmark between the implementations.

i think that is an incorrect assumption

**Comment by david415** [comment]
terrible no progress not sure how to fix it using the heap library.

**Comment by mixmasala** [comment]
> > I think this is probabalistically safe as our PriorityQueue may not handle items with the same priority appropriately; it would also be nice to see a comparison benchmark between the implementations.
> 
> i think that is an incorrect assumption

could I help clarify?

**Pull request closed without merging**